halo2_proofs/src/multicore.rs

```diff
@@ -1,9 +1,61 @@
 //! An interface for dealing with the kinds of parallel computations involved in
-//! `halo2`. It's currently just a (very!) thin wrapper around [`rayon`] but may
+//! `halo2`. It's currently just a (very!) thin wrapper around `rayon` but may
 //! be extended in the future to allow for various parallelism strategies.
 
-pub use rayon::{
-    current_num_threads,
-    iter::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator},
+pub use maybe_rayon::{
+    iter::{IntoParallelIterator, ParallelIterator},
     join, scope, Scope,
 };
+
+#[cfg(feature = "multicore")]
+pub use maybe_rayon::{current_num_threads, iter::IndexedParallelIterator};
+
+#[cfg(not(feature = "multicore"))]
+pub fn current_num_threads() -> usize {
+    1
+}
+
+#[cfg(not(feature = "multicore"))]
+pub trait IndexedParallelIterator: std::iter::Iterator {}
+
+pub trait TryFoldAndReduce<T, E> {
+    /// Implements `iter.try_fold().try_reduce()` for `rayon::iter::ParallelIterator`,
+    /// falling back on `Iterator::try_fold` when the `multicore` feature flag is
+    /// disabled.
+    /// try_fold_and_reduce function can only be called by a iter with
+    /// Result<T,E> item type because the folder_op Fn<T,T>->T must
+    /// meet both trait bound of rayon's try_fold and try_reduce```   
```

Comment:
```suggestion
    /// The `try_fold_and_reduce` function can only be called by a iter with
    /// `Result<T, E>` item type because the `fold_op: Fn<T, T> -> T` must
    /// meet both the trait bounds of rayon's `try_fold` and `try_reduce`.   
```